couchbase.py

```diff
@@ -0,0 +1,364 @@
+            if key == 'storageTotals':
+                metric_name_pref = 'storage'
+                metrics.extend(
+                        _parse_with_prefix(metric_name_pref, obj, dimensions,
+                                           module_config))
+            if key == 'nodes':
+                metric_name_pref = 'nodes'
+                for node in obj:
+                    node_dim = dimensions
+                    host = node.get('hostname')
+                    node_dim['node'] = host
+                    metrics.extend(
+                            _parse_with_prefix(metric_name_pref, node,
+                                               node_dim,
+                                               module_config))
+    elif request_type == REQUEST_TYPE_BUCKET:
+        for key, value in obj_to_parse.iteritems():
+            if key == 'quota':
+                metric_name_pref = 'bucket.quota'
+                metrics.extend(
+                        _parse_with_prefix(metric_name_pref, value, dimensions,
+                                           module_config))
+            if key == 'basicStats':
+                metric_name_pref = 'bucket.basic'
+                metrics.extend(
+                        _parse_with_prefix(metric_name_pref, value, dimensions,
+                                           module_config))
+    elif request_type == REQUEST_TYPE_BUCKET_STAT:
+        for key, value in obj_to_parse.iteritems():
+            if key == 'op':
+                samples = value.get('samples')
+                metric_name_pref = 'bucket.op'
+                for key_sample, value_sample in samples.iteritems():
+                    if isinstance(value_sample, list):
+                        metric_value = value_sample[-1]
+                        metrics.append(
+                                _process_metric(metric_name_pref, key_sample,
+                                                metric_value, dimensions,
+                                                module_config))
+                break
+
+    metrics = filter(lambda x: x is not None, metrics)
+    collectd.info("End parsing: " + str(len(metrics)))
+    for metric in metrics:
+        collectd.info(str(metric))
+    return metrics
+
+
+def _format_dimensions(dimensions):
+    """
+    Formats a dictionary of dimensions to a format that enables them to be
+    specified as key, value pairs in plugin_instance to signalfx. E.g.
+    dimensions = {'a': 'foo', 'b': 'bar'}
+    _format_dimensions(dimensions)
+    "[a=foo,b=bar]"
+    Args:
+    dimensions (dict): Mapping of {dimension_name: value, ...}
+    Returns:
+    str: Comma-separated list of dimensions
+    """
+    # Collectd limits the plugin_instance field size, so truncate anything
+    # longer than that.
+    trunc_len = field_length - 2  # account for the 2 brackets at either end
+    dim_pairs = []
+    # Put the name dimension first because it is more likely to be unique
```

Comment:
This was true for rabbitmq but might not be true for couchbase- you should check. The goal here was to put the dimension values that were mostly likely to change between metrics at the front of the string so they wouldn't get truncated.